In [1]:
import numpy as np
import time


In [2]:
# Function rosembrock to be optimized
def rosembrock_func(x):
    return 100 * (x[1][0] - x[0][0] ** 2) ** 2 + (1 - x[0][0]) ** 2


In [3]:
# Gradient of the function rosembrock to be optimized: delta f(x) = [df/dx[0], df/dx[1]] (2,1)
def grad_func_rosembrock(x):
    return np.array([-400 * x[0][0] * (x[1][0] - x[0][0] ** 2) - 2 * (1 - x[0][0]), 200 * (x[1][0] - x[0][0] ** 2)]).reshape(2,1)

In [4]:
x0 = np.array([[1.2], [1.2]])

In [5]:
# Hessian of the function rosembrock to be optimized: delta^2 f(x) = [[d^2f/dx[0]^2, d^2f/dx[0]dx[1], [d^2f/dx[1]dx[0], d^2f/dx[1]^2]] (2,2)
def hessian_rosembrock(x):
    return np.array([[1200 * x[0][0] ** 2 - 400 * x[1][0] + 2, -400 * x[0][0]], [-400 * x[0][0], 200]]).reshape(2,2)


In [6]:
print(x0)
print(rosembrock_func(x0))
print(grad_func_rosembrock(x0))
print(hessian_rosembrock(x0))

[[1.2]
 [1.2]]
5.8
[[115.6]
 [-48. ]]
[[1250. -480.]
 [-480.  200.]]


In [7]:
# Backtracking line search
def backtracking_Rosembrock(x0, f, grad_f, descent_direction, alpha=0.3, beta=0.8):
    x = x0
    t = 1
    while f(x + t* descent_direction) > f(x) + alpha * t * np.dot(grad_f(x).T, descent_direction):
        t *= beta
    return t

## 1 Gradient descent method with back tracking linesearch

In [8]:
# Gradient descent method with ∆x = - ∇f(x)
def gradient_descent(x0, f, grad_f, line_search, epsilon=1e-4, max_iter=1000):
    x = x0
    for i in range(max_iter+1):
        print(f"Epoch {i}: x = {x}, f(x) = {f(x)}")
        if np.linalg.norm(grad_f(x)) < epsilon:
            break
        descent_direction = -grad_f(x)
        t = line_search(x, f, grad_f, descent_direction)
        x += t * descent_direction
    return f(x)

In [9]:
print("First starting point: [1.2, 1.2]\n")
x0 = np.array([[1.2], [1.2]])
gradient_descent(x0, rosembrock_func, grad_func_rosembrock, backtracking_Rosembrock)


First starting point: [1.2, 1.2]

Epoch 0: x = [[1.2]
 [1.2]], f(x) = 5.8
Epoch 1: x = [[1.08551531]
 [1.2475369 ]], f(x) = 0.48608580118458417
Epoch 2: x = [[1.11510027]
 [1.23383173]], f(x) = 0.02249649673688547
Epoch 3: x = [[1.11062416]
 [1.23573655]], f(x) = 0.012744189875130976
Epoch 4: x = [[1.11139519]
 [1.23529078]], f(x) = 0.012409726600546187
Epoch 5: x = [[1.11111339]
 [1.23526246]], f(x) = 0.012393726131045347
Epoch 6: x = [[1.11124371]
 [1.23504908]], f(x) = 0.012378640942933571
Epoch 7: x = [[1.1109737 ]
 [1.23497693]], f(x) = 0.0123661941307079
Epoch 8: x = [[1.11112149]
 [1.23475584]], f(x) = 0.012350704684917745
Epoch 9: x = [[1.11089098]
 [1.23470482]], f(x) = 0.012336003421185121
Epoch 10: x = [[1.11100009]
 [1.23446263]], f(x) = 0.012323019985978852
Epoch 11: x = [[1.11075382]
 [1.23441886]], f(x) = 0.012307986559492279
Epoch 12: x = [[1.11087951]
 [1.23416941]], f(x) = 0.012295614937781899
Epoch 13: x = [[1.1106162 ]
 [1.23413347]], f(x) = 0.012280183541493262
Epo


Epoch 253: x = [[1.09611009]
 [1.20201581]], f(x) = 0.009268340042144697
Epoch 254: x = [[1.09621192]
 [1.20179976]], f(x) = 0.00925815411217088
Epoch 255: x = [[1.09599503]
 [1.20176287]], f(x) = 0.009246156326008581
Epoch 256: x = [[1.09609665]
 [1.2015471 ]], f(x) = 0.00923598729561772
Epoch 257: x = [[1.09588014]
 [1.20151019]], f(x) = 0.009224016539927736
Epoch 258: x = [[1.09598143]
 [1.20129475]], f(x) = 0.009213861430245367
Epoch 259: x = [[1.09576542]
 [1.20125778]], f(x) = 0.009201920665029483
Epoch 260: x = [[1.09586626]
 [1.20104271]], f(x) = 0.009191776555225739
Epoch 261: x = [[1.09565087]
 [1.20100562]], f(x) = 0.009179868704041055
Epoch 262: x = [[1.09575115]
 [1.20079099]], f(x) = 0.009169732742926586
Epoch 263: x = [[1.09553648]
 [1.20075373]], f(x) = 0.009157860680579588
Epoch 264: x = [[1.0956361 ]
 [1.20053958]], f(x) = 0.009147730097707252
Epoch 265: x = [[1.09542226]
 [1.2005021 ]], f(x) = 0.009135896638302691
Epoch 266: x = [[1.0955211 ]
 [1.20028849]], f(x) = 

0.003563270857269817

In [10]:
print("Second starting point: [-1.2, 1]\n")
x0 = np.array([[-1.2], [1]])
gradient_descent(x0, rosembrock_func, grad_func_rosembrock, backtracking_Rosembrock)


Second starting point: [-1.2, 1]

Epoch 0: x = [[-1.2]
 [ 1. ]], f(x) = 24.199999999999996
Epoch 1: x = [[-0.9864801 ]
 [ 1.08715098]], f(x) = 5.245885307456412
Epoch 2: x = [[-1.03725258]
 [ 1.05892397]], f(x) = 4.179192629995579
Epoch 3: x = [[-1.02349295]
 [ 1.06312528]], f(x) = 4.118820601752008
Epoch 4: x = [[-1.02710538]
 [ 1.05830119]], f(x) = 4.110282304828079
Epoch 5: x = [[0.68523434]
 [0.62876632]], f(x) = 2.6341851039714874
Epoch 6: x = [[0.77086651]
 [0.56717098]], f(x) = 0.12574924656979242
Epoch 7: x = [[0.75561105]
 [0.57764093]], f(x) = 0.06420540863127483
Epoch 8: x = [[0.76046931]
 [0.57505176]], f(x) = 0.058438896583951724
Epoch 9: x = [[0.75947675]
 [0.57631361]], f(x) = 0.05787557457493983
Epoch 10: x = [[0.76104357]
 [0.57677765]], f(x) = 0.05768082581184615
Epoch 11: x = [[0.76042549]
 [0.57794289]], f(x) = 0.05740518945189825
Epoch 12: x = [[0.76188629]
 [0.57817261]], f(x) = 0.05722626856636835
Epoch 13: x = [[0.76134438]
 [0.57928391]], f(x) = 0.0569695643093

Epoch 16: x = [[0.76357919]
 [0.5809314 ]], f(x) = 0.056344993948618916
Epoch 17: x = [[0.76304963]
 [0.58221394]], f(x) = 0.05614557291590941
Epoch 18: x = [[0.76445349]
 [0.58223255]], f(x) = 0.05594724458309807
Epoch 19: x = [[0.76399809]
 [0.58327542]], f(x) = 0.05571434693431749
Epoch 20: x = [[0.76529907]
 [0.58359099]], f(x) = 0.055522037093936105
Epoch 21: x = [[0.76478255]
 [0.58485533]], f(x) = 0.05532738426334584
Epoch 22: x = [[0.76617013]
 [0.58487771]], f(x) = 0.055133921450032555
Epoch 23: x = [[0.76571591]
 [0.58591205]], f(x) = 0.0549057483718387
Epoch 24: x = [[0.76701307]
 [0.58622093]], f(x) = 0.0547189343862394
Epoch 25: x = [[0.76648515]
 [0.58748312]], f(x) = 0.05452920979662087
Epoch 26: x = [[0.76760223]
 [0.58749104]], f(x) = 0.05430530174592227
Epoch 27: x = [[0.76736055]
 [0.58879225]], f(x) = 0.054121364807056614
Epoch 28: x = [[0.76872042]
 [0.58882245]], f(x) = 0.05393487875856655
Epoch 29: x = [[0.76827113]
 [0.58984212]], f(x) = 0.05371414129210599
Epoc

0.0046728427256420004

## 2 Newton method with back tracking line search

In [11]:
import numpy as np

x = np.array([[2, 3],
              [4, 5]])
y = np.linalg.inv(x)

result = np.dot(x, y)
print(result)


[[1. 0.]
 [0. 1.]]


In [12]:
# Newton method with ∆x = - (∇²f(x))⁻¹ ∇f(x)
def newton_method(x0, f, grad_f, hessian, line_search, epsilon=1e-4, max_iter=1000):
    x = x0
    for i in range(max_iter+1):
        print(f"Epoch {i}: x = {x}, f(x) = {f(x)}")
        # Stopping criteria
        # norm of delta < epsilon
        if np.linalg.norm(grad_f(x)) < epsilon:
            break
        # lambda2 / 2 <= epsilon with lambda2 = ∇f(x)ᵀ (∇²f(x))⁻¹ ∇f(x)
        lambda2 = np.dot(grad_f(x).T, np.dot(np.linalg.inv(hessian(x)), grad_f(x)))
        if lambda2 / 2 <= epsilon:
            break
        descent_direction = -np.dot(np.linalg.inv(hessian(x)), grad_f(x))
        
        t = line_search(x, f, grad_f, descent_direction)
        x += t * descent_direction
    return f(x)

In [13]:
print("First starting point: [1.2, 1.2]\n")
x0 = np.array([[1.2], [1.2]])
newton_method(x0, rosembrock_func, grad_func_rosembrock, hessian_rosembrock, backtracking_Rosembrock)


First starting point: [1.2, 1.2]

Epoch 0: x = [[1.2]
 [1.2]], f(x) = 5.8
Epoch 1: x = [[1.19591837]
 [1.43020408]], f(x) = 0.038384034418534184
Epoch 2: x = [[1.09594128]
 [1.19108374]], f(x) = 0.01921182605186693
Epoch 3: x = [[1.06396842]
 [1.13100653]], f(x) = 0.004196460661754682
Epoch 4: x = [[1.01085848]
 [1.01901419]], f(x) = 0.0009135219903405465
Epoch 5: x = [[1.00391631]
 [1.00779976]], f(x) = 1.55697292474431e-05


1.55697292474431e-05

In [14]:
print("Second starting point: [-1.2, 1]\n")
x0 = np.array([[-1.2], [1]])
newton_method(x0, rosembrock_func, grad_func_rosembrock, hessian_rosembrock, backtracking_Rosembrock)

Second starting point: [-1.2, 1]

Epoch 0: x = [[-1.2]
 [ 1. ]], f(x) = 24.199999999999996
Epoch 1: x = [[-1.1752809 ]
 [ 1.38067416]], f(x) = 4.731884325266608
Epoch 2: x = [[-0.91511382]
 [ 0.76921738]], f(x) = 4.13300226320981
Epoch 3: x = [[-0.7843285 ]
 [ 0.59806639]], f(x) = 3.2130856128676224
Epoch 4: x = [[-0.52602031]
 [ 0.20381651]], f(x) = 2.8598998169973844
Epoch 5: x = [[-0.42804883]
 [ 0.1736274 ]], f(x) = 2.048536419892932
Epoch 6: x = [[-0.22770892]
 [ 0.00604837]], f(x) = 1.7170605076532184
Epoch 7: x = [[-0.10687852]
 [-0.00317697]], f(x) = 1.2464960184633609
Epoch 8: x = [[ 0.11901522]
 [-0.03978336]], f(x) = 1.0671726284538807
Epoch 9: x = [[0.19374083]
 [0.03195159]], f(x) = 0.6531718577280331
Epoch 10: x = [[0.38875591]
 [0.11037533]], f(x) = 0.5397231243685873
Epoch 11: x = [[0.45555003]
 [0.20306437]], f(x) = 0.29841622839711246
Epoch 12: x = [[0.60286268]
 [0.33956521]], f(x) = 0.21473492952674286
Epoch 13: x = [[0.67162341]
 [0.44634997]], f(x) = 0.11006661799

4.4920548861222697e-07

## 3 Compare the convergence speed of the two algorithms.

We can see that the first method - gradient descent costs us 1000 iterations or more to get the answer for the minimization of the function:


x = [[1.2]
 [1.2]] <br>
Epoch 1000: x = [[1.05972113]
 [1.12312324]], f(x) = 0.003567921126687234

 x = [[-1,2]
 [1]] <br>
Epoch 1000: x = [[0.93171918]
 [0.86765623]], f(x) = 0.004682019824241827

Otherwise, the second method - newton_method - is more efficient than the first one - gradient_descent - because it converges in less iterations and get more efficient answer as you can see below:

x = [[1.2]
 [1.2]] <br>
Epoch 5: x = [[1.00391631]
 [1.00779976]], f(x) = 1.55697292474431e-05

x = [[-1,2]
 [1]] <br>
Epoch 19: x = [[0.99933347]
 [0.99866035]], f(x) = 4.4920548861222697e-07